service/history/ndc/new_workflow_transaction_manager.go

```diff
@@ -157,10 +157,20 @@ func (r *transactionManagerForNewWorkflowImpl) createAsCurrent(
 		return err
 	}
 
-	targetWorkflowHistorySize, err := targetWorkflow.GetContext().PersistFirstWorkflowEvents(
-		ctx,
-		targetWorkflowEventsSeq[0],
-	)
+	var targetWorkflowHistorySize int64
+	if len(targetWorkflowEventsSeq[0].Events) > 0 {
+		if targetWorkflowEventsSeq[0].Events[0].GetEventType() == types.EventTypeWorkflowExecutionStarted {
+			targetWorkflowHistorySize, err = targetWorkflow.GetContext().PersistStartWorkflowBatchEvents(
+				ctx,
+				targetWorkflowEventsSeq[0],
+			)
+		} else { // reset workflows fall into else branch
+			targetWorkflowHistorySize, err = targetWorkflow.GetContext().PersistNonStartWorkflowBatchEvents(
+				ctx,
+				targetWorkflowEventsSeq[0],
+			)
+		}
+	}
```

Comment:
can you provide please some context here?

Reply:
for reset workflows, the first event is not workflow started, but decisiontaskfailed or decisiontasktimedout. and the history of reset workflows are created from a forked branch of an existing workflows and they don't require creating a new branch.

Reply:
Just a note that this bug should only exists for sql because Cassandra uses upsert command for inserting

Reply:
makes sense, thanks @longquanzheng!